src/Driver/PDO/Statement.php

```diff
@@ -117,12 +133,109 @@ public function execute($params = null): ResultInterface
             );
         }
 
+        $resourceOffsets = $this->getResourceOffsets();
         try {
             $this->stmt->execute($params);
         } catch (PDOException $exception) {
             throw Exception::new($exception);
+        } finally {
+            if ($resourceOffsets !== null) {
+                $this->restoreResourceOffsets($resourceOffsets);
+            }
         }
 
         return new Result($this->stmt);
     }
+
+    /**
+     * Track a binary parameter reference at binding time. These
+     * are cached for later analysis by the getResourceOffsets.
+     *
+     * @param mixed $resource
+     */
+    private function trackParamResource($resource): void
+    {
+        if (! is_resource($resource)) {
+            return;
+        }
+
+        $resources   = &$this->paramResources;
+        $resources ??= [];
+        $resources[] = $resource;
+    }
+
+    /**
+     * Determine the offset that any resource parameters needs to be
+     * restored to after the statement is executed. Call immediately
+     * before execute (not during bindValue) to get the most accurate offset.
+     *
+     * @return int[]|null Return offsets to restore if needed.
+     */
+    private function getResourceOffsets()
+    {
+        $resources = &$this->paramResources;
+        if ($resources === null) {
+            return null;
+        }
+
+        $count           = count($resources);
+        $resourceOffsets = null;
+        $deref           = [];
+        for ($i = 0; $i < $count; ++$i) {
+            $resource = $resources[$i];
+            $position = false;
+            try {
+                if (stream_get_meta_data($resource)['seekable']) {
+                    $position = ftell($resource);
+                }
+            } catch (Throwable $e) {
+                // Swallow
+            } finally {
+                if ($position === false) {
+                    if ($resourceOffsets !== null) {
+                        $resourceOffsets[] = -1;
+                    }
+                } else {
+                    $resourceOffsets ??= $i !== 0 ? array_fill(0, $i, -1) : [];
```

Comment:
What is the point of this `array_fill()` operation?

Reply:
The $resourceOffsets array is not initialized until it encounters the first restorable offset, which can be in a position after 0. The -1 special number is used in the following restoreResourceOffsets loops to skip the value.

Earlier iterations of this code did not add a resource to the $paramResources array unless a restore offset could be obtained. However, this meant retrieving this information in bindValue, which leaves an indeterminate amount of time between when the resource was bound and when it was used in execute. I move the offset retrieval immediately before execute to narrow this gap, but that means the previous 'don't track a non-resettable-resource' approach was no longer feasible. Instead of (possibly repeatedly) splicing the $paramResources array in this situation I just kept the same array size and marked the resourceOffsets with a special 'skip' value.

Reply:
I doubt that a sparse array will cause trouble here. An array that only holds the offsets that we need is just fine.

Reply:
This has been moved to a sparse array. The index is still needed, so foreach iteration is used to extract both the key and value when it is used later on.